homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -615,15 +593,15 @@ class XiaomiAirPurifierMiot(XiaomiAirPurifier):
     @property
     def percentage(self):
         """Return the current percentage based speed."""
-        if self._state:
+        if self.is_on:
             return ranged_value_to_percentage((1, 3), self._fan_level)
 
         return None
 
     @property
     def preset_mode(self):
         """Get the active preset mode."""
-        if self._state:
+        if self.is_on:
```

Comment:
I got an exception here:
```
Traceback (most recent call last):
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity_platform.py", line 382, in async_add_entities
    await asyncio.gather(*tasks)
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity_platform.py", line 587, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity.py", line 699, in add_to_platform_finish
    self.async_write_ha_state()
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity.py", line 464, in async_write_ha_state
    self._async_write_ha_state()
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity.py", line 498, in _async_write_ha_state
    state = self._stringify_state()
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity.py", line 470, in _stringify_state
    state = self.state
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity.py", line 883, in state
    return STATE_ON if self.is_on else STATE_OFF
  File "/home/maciek/develop/home-assistant-core/homeassistant/components/fan/__init__.py", line 435, in is_on
    return self.speed not in [SPEED_OFF, None]
  File "/home/maciek/develop/home-assistant-core/homeassistant/components/fan/__init__.py", line 462, in speed
    preset_mode = self.preset_mode
  File "/home/maciek/develop/home-assistant-core/homeassistant/components/xiaomi_miio/fan.py", line 604, in preset_mode
    if self.is_on:
...
...
...
  File "/home/maciek/develop/home-assistant-core/homeassistant/components/fan/__init__.py", line 447, in _implemented_preset_mode
    return not hasattr(self.set_preset_mode, _FAN_NATIVE) or not hasattr(
  RecursionError: maximum recursion depth exceeded while calling a Python object


```

Reply:
Maybe `if self.coordinator.data.is_on:`?

Reply:
Hmm. Looks like `is_on` from the parent calls `preset_mode` whose implementation again checks for `is_on`, which is causing this. I'll need to dig into the code more to see what is the proper fix. `_attr_is_on` is set to `self.coordinator.data.is_on` in the update callback :-P